### PR TITLE
[10.0][ADD] shopinvader_delivery_carrier_category_keep_carrier

### PR DESCRIPTION
**Pull request by rousseldenis**

### CODE REVIEW

**Comment by rousseldenis** [comment]
@Cedric-Pigeon

**Review by Cedric-Pigeon — APPROVED** [review]
LGTM (code review)

**Comment by rousseldenis** [comment]
@sebastienbeau @lmignon

File: shopinvader_delivery_carrier_category_keep_carrier/services/cart.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+from odoo.addons.component.core import Component
+
+_logger = logging.getLogger(__name__)
+
+
+class CartService(Component):
+
+    _inherit = "shopinvader.cart.service"
+
+    def _check_carrier_unset(self, cart, params):
```
Comment (lmignon) [inline shopinvader_delivery_carrier_category_keep_carrier/services/cart.py:15]: @rousseldenis The method name is a little bit confusing or at least lack of meaning. (The same applies to the docstring) The purpose of this method is to know if the carrier should be reset on the cat when the shipping address change on the SO... `_reset_carrier_on_shipping_address_update`

File: shopinvader_delivery_carrier_category_keep_carrier/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# -*- coding: utf-8 -*-
+# Copyright 2020 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Shopinvader Delivery Carrier Category Keep Carrier",
+    "summary": """
+        Allows to keep carrier even on delivery address change""",
```
Comment (lmignon) [inline shopinvader_delivery_carrier_category_keep_carrier/__manifest__.py:8]: Where is the `readme` directory? Can you provide at least a description on how this addon works? At least, we should find into the description that this addon allows you to specify on a delivery carrier category if the carrier information on the SO must be reset when the shipping address change.

**Review by lmignon — CHANGES_REQUESTED** [review]
@rousseldenis IMO this functionality should go into a addon outside shopinvader. (No business logic into shopinvader) This behaviour is 'transversal'

File: shopinvader_delivery_carrier_category_keep_carrier/services/cart.py (hunk 1 above)
Comment (rousseldenis) [reply shopinvader_delivery_carrier_category_keep_carrier/services/cart.py:15]: > @rousseldenis The method name is a little bit confusing or at least lack of meaning. (The same applies to the docstring) The purpose of this method is to know if the carrier should be reset on the cat when the shipping address change on the SO... `_reset_carrier_on_shipping_address_update`

It's a behaviour typically done in shopinvader when changing the shipping adress. So, a module outside shopinvader would have no utility. This method is a simple hook I've added but could have been done in _unset_carrier()

**Comment by rousseldenis** [comment]
@lmignon Could you update ?

File: shopinvader_delivery_carrier/services/cart.py [hunk 3]
```diff
@@ -112,9 +112,13 @@ def _delete_item(self, cart, params):
         self._unset_carrier(cart)
         return res
 
+    def _check_carrier_unset(self, cart, params):
+        return True
+
```
Comment (lmignon) [inline shopinvader_delivery_carrier/services/cart.py:117]:
```suggestion
    def _must_unset_carrier(self, cart, params):
        return params.get("shipping", {}).get("address"):
```

File: shopinvader_delivery_carrier/services/cart.py [hunk 4]
```diff
@@ -112,9 +112,13 @@ def _delete_item(self, cart, params):
...
     def _update(self, cart, params):
         if params.get("shipping", {}).get("address"):
```
Comment (lmignon) [inline shopinvader_delivery_carrier/services/cart.py:119]:
```suggestion
    def _update(self, cart, params):
```

File: shopinvader_delivery_carrier/services/cart.py [hunk 5]
```diff
@@ -112,9 +112,13 @@ def _delete_item(self, cart, params):
...
-            self._unset_carrier(cart)
+            if self._check_carrier_unset(cart, params):
+                self._unset_carrier(cart)
```
Comment (lmignon) [inline shopinvader_delivery_carrier/services/cart.py:121]:
```suggestion
        if self._must_unset_carrier(cart, params):
            self._unset_carrier(cart)
```

File: shopinvader_delivery_carrier_category_keep_carrier/services/cart.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+        """
+        The carrier category does not need to unset
+        :param cart:
+        :param params:
+        :return:
+        """
+        if cart.carrier_id.category_id.keep_carrier_on_shipping_change:
+            return False
+        return super(CartService, self)._check_carrier_unset(cart, params)
```
Comment (lmignon) [inline shopinvader_delivery_carrier_category_keep_carrier/services/cart.py:24]:
```suggestion
    def _must_unset_carrier(self, cart, params):
        """
        The carrier category does not need to unset
        :param cart:
        :param params:
        :return:
        """
        must_unset_carrier = super(CartService, self)._check_carrier_unset(cart, params)
        return  must_unset_carrier and not cart.carrier_id.category_id.keep_carrier_on_shipping_change
```

**Review by lmignon — CHANGES_REQUESTED** [review]

**Comment by lmignon** [comment]
Closing since 10.0 is no more actively maintained.

**Pull request closed without merging**